_posts/2019/2019-10-10-building-a-drawer-component-in-react.md

```diff
@@ -0,0 +1,805 @@
+
+``` scss
+.Drawer {
+  &-title {
+    background: #efefef;
+    border: 1px solid;
+    cursor: pointer;
+    display: inline-block;
+    padding: 16px;
+  }
+
+  &-content {
+    ...
+  }
+}
+```
+
+The result is a primitive looking button to click. Which brings up something else we should probably fix at this point which is keyboard accessibility. Time for a slight detour.
+
+![Basic title styles](/img/2019/basic-title-style.png)
+
+### Keyboard Accessibility
+
+Right now on the page anyone using a screen reader or other non-mouse assistance technology, will be unable to tab to our new button. One way we could attempt to fix this is by giving our `Drawer-title` div a `tabIndex` value and a `role` value.
+
+``` jsx
+<div
+  className="Drawer-title"
+  onClick={this.handleTitleClick}
+  tabIndex="0"
+  role="button"
+>
+  {title}
+</div>
+```
+
+<blockquote class="Info Info-right">
+  The eslint plugin jsx-a11y has some great documentation on the subject of non-semantic
+  elements having an onClick handler.
+  <cite><a href="https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/click-events-have-key-events.md" target="_blank">ESlint jsx-a11y</a></cite>
+</blockquote>
+
+While, this allows us to focus on the element, the enter and spacebar keys don't activate our `onClick` handler. Major bummer. We could try to add an `onKeyPress` event and capture the spacebar and enter key, but why do more than we need to when there is a perfectly semantic approach. The `<button>` element.
+
+So let's fix this back in our `render()` method.
+
+``` jsx
+render() {
+  const { title, children } = this.props;
+  const drawerStyles = this.state.active ? 'is-expanded' : '';
+
+  return (
+    <div className={`Drawer ${drawerStyles}`}>
+      <button
+        className="Drawer-title"
+        onClick={this.handleTitleClick}
+      >
+        {title}
+      </button>
+      <div className="Drawer-content">
+        {children}
+      </div>
+    </div>
+  );
+}
+```
+
+As you can see from the above, we no longer need `tabIndex`, `role`, or even additonal events like `onKeyPress`. We can also remove the `Drawer-title` style for `display: inline-block` since buttons by default are inline-block.
+
+At this point our button just works exactly how we would expect. Hitting the tab key on the page will focus on our new button. Once focussed pressing either enter or spacebar will activate our `onClick` handler.
+
+### Hiding and showing content
+
+Now that we've fixed our button's accesibility issues we can focus back on our goal of showing and hiding content in the Drawer. Since the height of the element starts out as 0 we can simply have its height change to a larger value when the `is-expanded` class is active via the `this.state.active` variable. We want this value to be greater than the containing content's height could ever be. I choose `40vh` here as it is 40% of the current viewports height (and it seemed a reasonable starting place).
```

Comment:
*give more of a transition here, reference the last piece of code you handled* Since the height of the element starts out as 0 we can simply have its height change to a larger value when the `is-expanded` class is active via the `this.state.active` variable. We want this value to be greater than the containing content's height could ever be. I choose `40vh` here as it is 40% of the current viewports height (and it seemed a reasonable starting place).